Add a generic mechanism for passing options

lhuard1A opened the pull request:
And use it in the libvirt playbook

The idea is to provide a generic way to customize the behavior of openshift-ansible playbooks.

1. `bin/cluster` learnt the `--option <key>=<value>` command line parameter. This parameter can be specified several times
2. For each `--option <key>=<value>` parameter, `bin/cluster` creates a `opt_<key>` ansible variable. The variable is prefixed with `opt_` to avoid potential name conflicts and to prevent the end user from overriding ansible variables which were not intended to be set by him.
3. Inside the playbooks, the options can be read by the `oo_option` lookup plugin. For ex. `lookup('oo_option', '<key> default=<default_value>')` returns, by order of priority:
  1. if it exists, the `opt_<key>` ansible variable. This variable is set by `bin/cluster --option <key>=<value> …`
  2. if it exists, the envirnoment variable named `<key>` (for backward compatibility with how current AWS and GCE playbooks can be customized)
  3. if present, the `<default_value>` passed in the lookup statement
  4. if none of the above conditions are met, empty string is returned

lhuard1A commented:
cc @twiest

twiest commented on lookup_plugins/oo_option.py:
```diff
@@ -0,0 +1,78 @@
+Usage:
+
+    - debug:
+      msg: "{{ lookup('oo_option', '<key> default=<default_value>') }}"
```
So why does the lookup plugin allow for defining a default, instead of using the built in default filter?

That would make it more consistent with our playbooks, more modular, and would allow us to use the features of the default filter (like replacing an empty string, empty list, etc with the default value if we want to).

For example:
```
msg: "{{ lookup('oo_option', '<key>') | default('<default_value>') }}"
```

detiber replied:
@twiest I'm not sure that would be sufficient, since the lookup plugin is testing multiple variables...  I think the proper solution will be to keep the default parameter and adding a boolean parameter to provide the same behavior as the default filter.

Examples:
- invalid env variable value overriding default
  - Playbook:

    ```
    name:   "{{ lookup('oo_option', 'image_name') |  default('CentOS-7-x86_64-GenericCloud.qcow2', true) }}"
    ```
  - Command invocation:

    ```
    export IMAGE_NAME=''
    bin/cluster <blah>
    ```
  - Would result in image_name being set to an empty string
- invalid -o option overriding valid env variable
  - Playbook:

    ```
    name:   "{{ lookup('oo_option', 'image_name default=CentOS-7-x86_64-GenericCloud.qcow2') }}"
    ```
  - Command invocation:

    ```
    export IMAGE_NAME=CentOS-7-x86_64-Custom.qcow2
    bin/cluster <blah> -o 'image_name=""'
    ```
  - Would result in image_name being set to an empty string

twiest replied:
@detiber I must be missing something. In the first example, wouldn't 'name' be set to an empty string?

And isn't that exactly the behavior we'd be looking for?

My understanding is that inside of the filter pipeline, all filters get the data and have a chance to mutate it and pass it along.

So, in the first example, the lookup plugin looks up the value (in this case empty string), and passes that to the default filter.

Then default filter takes that value and determines if it should change it. In this case, since we passed in 'true' as the second parameter, empty strings do get replaced.

So, wouldn't 'name' end up being the value 'CentOS-7-x86_64-GenericCloud.qcow2' ?

I agree that in the second example, 'name' would be set to an empty string, but again, that's what we'd expect. I think I'm totally missing your point.

detiber replied:
> @detiber I must be missing something. In the first example, wouldn't 'name' end up being the value 'CentOS-7-x86_64-GenericCloud.qcow2' ?
> 
> And isn't that exactly the behavior we'd be looking for?

Yes, not sure what I was thinking there, I must have mixed up my example and I can't think of what I was trying to convey now.

> Then default filter takes that value and determines if it should change it. In this case, since we passed in 'true' as the second parameter, empty strings do get replaced.
> 
> So, again, wouldn't 'name' end up being the value 'CentOS-7-x86_64-GenericCloud.qcow2' ?
> 
> I agree that in the second example, 'name' would be set to an empty string, but again, that's what we'd expect. I think I'm totally missing your point.

In the second example, the default behavior isn't carried through the possible defaults, where we want to treat the value as a python boolean, so the flag setting the value to "" should be treated as a python False and we should instead use the env set default.

twiest replied:
@detiber Ah, ok, I see what you're saying now.

I still don't think we want lookup to have a 'default' to set, though.

Instead, I think we want the lookup plugin to have the same param as the default filter, where we can tell it to skip values that evaluate to false.

Something like this:
```
# Before
"{{ lookup('oo_option', 'image_name   default=CentOS-7-x86_64-GenericCloud.qcow2') }}"

# My proposal
"{{ lookup('oo_option', 'image_name   skip_empty=true') | default('CentOS-7-x86_64-GenericCloud.qcow2', true) }}"
```

Thoughts?

detiber replied:
@twiest Between this and the latter comment regarding returning None for non-existent keys, I think that is a good approach.

So, removing the idea of having a default or even needing to use the boolean param for the default filter, will provide consistent behavior across the board, so if a variable is set on the command line it will override anything set in the env and either will override the default provided by the default filter.

lhuard1A replied:
Hello,
Look at my last comment. I have the feeling that a lookup plugin which fails to find what it is looking for is supposed to return an empty string.

As a consequence, if `FOO` is not set
```yaml
  - debug: msg="Env FOO={{ lookup('env', 'FOO') | default('default filter value') }}"
```
returns:
```
TASK: [debug msg="Env FOO="] ************************************************** 
ok: [localhost] => {
    "msg": "Env FOO="
}
```

if we set the boolean to True:
```yaml
  - debug: msg="Env FOO={{ lookup('env', 'FOO') | default('default filter value', True) }}"
```
we get the default value as expected:
```
TASK: [debug msg="Env FOO=default filter value"] ****************************** 
ok: [localhost] => {
    "msg": "Env FOO=default filter value"
}
```

The problem is that I cannot get empty string anymore because if I set the environment variable to `""` (empty string), it will be replaced by the `default` filter.

In short, it is not possible to distinguish between a lookup failure and a successful lookup that found a key the value of which is empty.
As a consequence, if we have
```yaml
lookup('…') | default('my_default_value', True)
```
It is not possible to get `''` empty string anymore.

If we have the default value handled by the `oo_option` plugin, it is possible to override a non-empty default value by setting the environment variable to empty string.

What are your feelings?

twiest replied:
@lhuard1A Thank you for researching this and for looking up the behavior of other ansible lookup modules (in this case 'env').

I agree that this limitation is less then ideal. I'd prefer them to let us return None (seems reasonable, right?). But since they don't seem to right now, I prefer our code to behave consistently with the ansible code. In that way, our users will know what to expect.

I think the corner case where we actually want an empty string can be handled in other ways (like we'd handle them for any of the built in lookup plugins).

I'd prefer that we don't add the default option to the oo_option plugin, and instead use the default filter.

This makes our plugin behave consistently with the built in ansible lookup plugins.

@detiber what are your thoughts based on this new data?

@lhuard1A we can discuss this further in the ansible architecture meeting next week if you feel strongly about it.

detiber replied:
@twiest :+1: on using the default filter.

twiest commented:
Hey Lenaic, thanks for submitting this PR, it looks really good. I let 1 comment.

@detiber What are your thoughts?

detiber commented:
@twiest lgtm outside of the one nit with the oo_option lookup plugin.

twiest commented on lookup_plugins/oo_option.py:
```diff
@@ -0,0 +1,78 @@
+                raise errors.AnsibleError(ex)
+
+            opt_key = 'opt_' + option_name
+            if opt_key in inject:
```
@lhuard1A sorry, just saw this.

Could inject be None here?

If so, I would prefer this:
```
if inject and opt_key in inject:
```

twiest commented on lookup_plugins/oo_option.py:
```diff
@@ -0,0 +1,78 @@
+            option_name = params[0]
+
+            paramvals = {
+                'default': ''
```
We probably want the lookup plugin to return a python 'None' if it doesn't find a value instead of empty string.

Empty string seems like the lookup plugin found the value and it was assigned an empty string.

lhuard1A replied:
I made a try to have the lookup plugin return `None` if the value cannot be found.

It doesn’t work.
With
```python
            paramvals = {
                'default': None
```
in `oo_option.py`, the following Ansible task:
```yaml
  - debug: msg="Var toto={{ lookup('oo_option', 'toto') | default('true default') }}"
```
fails with:
```
TASK: [debug msg="Var toto={{ lookup('oo_option', 'toto') | default('true default') }}"] *** 
fatal: [localhost] => Failed to template msg="Var toto={{ lookup('oo_option', 'toto') | default('true default') }}": an unexpected type error occurred. Error was sequence item 0: expected string, NoneType found
```

I looked at how the “standard” lookup plugins are written and it seems they return an empty string:

* https://github.com/ansible/ansible/blob/devel/v1/ansible/runner/lookup_plugins/env.py#L40
* https://github.com/ansible/ansible/blob/devel/v1/ansible/runner/lookup_plugins/redis_kv.py#L67

When the lookup fails, it is `''` which is returned.

twiest replied:
This is very interesting, thanks for researching this.

lhuard1A commented:
@twiest, @detiber, I have
* updated the `oo_option` lookup plugin to make it behave like other “regular” lookup plugins (no default feature embedded anymore, in case of lookup failure, it returns empty string and the regular default filter can be used behind)
* rebased
* squashed

twiest commented:
@lhuard1A great job with this!

Sorry this took so long to merge, but I really think this is a great implementation.

:+1:
